### PR TITLE
Interpolate predictions

### LINKED ISSUE #46: Interpolate predictions to observation depths
The direct outputs of the ML model are predicted temperatures at many fixed depths per day. Let's interpolate those predictions to the depths of observations for the sake of model evaluation, and for consistency with lake-temperature-process-models (see [this issue](https://github.com/USGS-R/lake-temperature-process-models/issues/59) for discussion and details).

Do this for the train, validation, and test sets.

### PR DESCRIPTION
**Pull request by AndyMcAliley**
Interpolate predicted temperatures from fixed depths to depths of observed temperatures.

The LSTM predicts temperatures at a set of fixed depths at every time step (day). Here we interpolate those predictions to the depths of every observation in a training, validation, or test set for the sake of model evaluation.

The trickiest part is to match each observation to the right prediction sequence. Training, validation, and testing sequences are formed as follows. Every sequence is a fixed number of days (set to 400 days right now). The first part of the sequence is "spin-up" or "burn-in" time (set to 100 days right now). Only predictions after spin-up are used for model evaluation. Sequences can overlap (set to 200 days right now). Only sequences that have an observation occur after spin-up are included in a dataset. Every observation for which we have NLDAS driver data has at least one corresponding sequence. Because of sequence overlap, some observations have more than one corresponding sequence.

When it comes to finding the right sequence for each observation, performance matters. My original implementation was a set of nested for loops over all observations and all sequences. It was on track to take 20 hours for the validation set. The current implementation takes 2-3 minutes. I'm using a combination of batch processing, [broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html), [np.nonzero](https://numpy.org/doc/stable/reference/generated/numpy.nonzero.html), and [np.where](https://numpy.org/doc/stable/reference/generated/numpy.where.html) to take advantage of vectorized operations.

The question of how to evaluate model performance was first raised in lake-temperature-process-models, [here](https://github.com/USGS-R/lake-temperature-process-models/issues/59). With this PR we can evaluate models in the same way as lake-temperature-process-models.

# How to run the code
The rule `interpolate_predictions` is the new addition. So, to run this part of the pipeline:
```
snakemake -c1 -s 4_evaluate.smk 4_evaluate/out/model_prep/initial/cpu_a/interpolated_predictions_valid.csv
```
That should make `interpolated_predictions_valid.csv` with the column `predicted_temperature_obs_depth` that contains the interpolated predictions.

> NOTE: Running the command above requires a compiled dataset and a trained model each of which takes several hours to obtain, plus a set of observed temperatures. If you want, copy the following files from Tallgrass, preserving their directory structure relative to the repo root directory:
> - 2_process/out/model_prep/valid.npz
> - 2_process/tmp/model_prep/temperature_observations_interpolated.csv
> - 3_train/out/model_prep/initial/cpu_a_metadata.npz
> - 3_train/out/model_prep/initial/cpu_a_weights.npz

# How to review this PR

The main things I'd like reviewed are:

1. Do the pipeline additions and modifications make sense?
2. Have any errors been introduced into the pipeline?

Closes #46.

### CODE REVIEW

File: 4_evaluate/src/interpolate_predictions.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                                 obs_file,
+                                 interpolated_predictions_filepath,
+                                 batch_size=5000):
+    """
+    Interpolate ML output to observation depths.
+
+    :param predictions_filepath: Path to unnormalized predictions for one
+        dataset (train, validation, or test)
+    :param metadata_filepath: Path to model training metadata npz for one dataset
+    :param dataset_filepath: Path to input for one dataset
```
Comment (jdiaz4302) [inline 4_evaluate/src/interpolate_predictions.py:53]: "path to input for one dataset" is vague/not as informative as the other lines of documentation

File: 4_evaluate/src/interpolate_predictions.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    # Load metadata to get depths_use
+    metadata_npz = np.load(metadata_filepath)
+    depths_use = metadata_npz['depths_use']
+
+    # Now, we'll need obs for each of those lakes.
+    # Read temperature observations
+    obs_all = pd.read_csv(obs_file, index_col=0, parse_dates=['date'])
+    # Whittle down to lakes in this dataset
+    obs = obs_all[obs_all.site_id.isin(site_ids)]
+    obs_all.site_id.unique().shape, obs.site_id.unique().shape, np.unique(site_ids).shape
```
Comment (jdiaz4302) [inline 4_evaluate/src/interpolate_predictions.py:79]: Does this actually print from snakemake / is it useful or is this a development relic?

If it is a development aid, perhaps an `assert(shape1 == shape2)` (like those you have later) would be a useful test to have here instead.

File: 4_evaluate/src/interpolate_predictions.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            # Select the sequence with the earliest start date
+            batch_start_dates = pred_starts_days[idx_sequence]
+            # Create a 2D array with the matches along rows and the observations in
+            # this batch along columns. Values are start dates where a match
+            # corresponds to an observation in this batch, and NaN otherwise.
+            start_date_matrix = np.where(
+                idx_obs[:, np.newaxis] == np.arange(this_batch_size),
+                batch_start_dates[:, np.newaxis],
+                np.NaN
+            )
```
Comment (jdiaz4302) [inline 4_evaluate/src/interpolate_predictions.py:144]: It seems to me that `idx_obs` is expected to be a shape less than `this_batch_size` due to the triple condition subsetting/index identification that happened earlier (Line 112). Is that intuition right? It seems that this would raise an error if that were the case.

For example, if the `batch_size` is 5 and `idx_obs` identifies the first observation as failing to meet the earlier conditions, we would get `idx_obs = [1, 2, 3, 4]` (right?) which would then error out when compared to `np.arange(5)`

File: 4_evaluate/src/interpolate_predictions.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    # Good. That means that the depth_below_index doesn't need to be used when it's one greater than max
+    # because those values get multiplied by zero
+    # Set those indices to the max index instead
+    depth_below_index[too_low] = len(depths_use) - 1
+
+    # Let's now skip any obs with indices equal to -1
+    # TODO: Consider dropping observations that are deeper than deepest depth_use
+    use_obs = ((sequence_indices >= 0) & (date_indices >= 0))
+
+    # Now the tricky part (like it hasn't been tricky so far!):
```
Comment (jdiaz4302) [inline 4_evaluate/src/interpolate_predictions.py:199]: Yes 😅. This looks like it was a lot of work that is tricky to think-up from scratch

**Review by jdiaz4302 — COMMENTED** [review]
_🤦🏼‍♂️ I still don't have this pipeline runnable because the large files failed to download off Tallgrass a couple times and I moved on to other things, but I will give that a go again_

Despite not running the pipeline, I mocked up a few minimal example tests of a lot of this functionality (e.g., the triple-condition index identification, the depth matching, the interpolation weight determination). All those minimal tests behaved as expected and the flow of operations  makes sense and seems efficient,

Specifically, I had one misunderstanding or concern regarding the `state_date_matrix` that I'd like addressed before approval, but other than looks great and like hard work

File: 4_evaluate/src/interpolate_predictions.py (hunk 1 above)
Comment (AndyMcAliley) [reply 4_evaluate/src/interpolate_predictions.py:53]: Agreed - reworded

File: 4_evaluate/src/interpolate_predictions.py (hunk 2 above)
Comment (AndyMcAliley) [reply 4_evaluate/src/interpolate_predictions.py:79]: Whoops! Yes, this is a pure dev relic. I think I'll leave out asserts for now

File: 4_evaluate/src/interpolate_predictions.py (hunk 3 above)
Comment (AndyMcAliley) [reply 4_evaluate/src/interpolate_predictions.py:144]: Your intuition is right: the length of `idx_obs` is generally not `this_batch_size`. `idx_obs` can be a different shape and this will work thanks to the broadcasting that's happening. If it were `idx_obs == np.arange(this_batch_size)` then it would error. To rephrase from the [numpy docs](https://numpy.org/doc/stable/user/basics.broadcasting.html#broadcastable-arrays):
> Here the `newaxis` index operator inserts a new axis into `idx_obs`, making it a two-dimensional nx1 array. Comparing the nx1 array with `np_arange(this_batch_size)`, which has shape `(this_batch_size,)`, yields a nx`this_batch_size` array.

It reminds me of an [outer product](https://en.wikipedia.org/wiki/Outer_product) in that one vector gets transposed and the result is a matrix where every element (i, j) is a function of the ith element of one vector and the jth element of the other.

File: 4_evaluate/src/interpolate_predictions.py (hunk 4 above)
Comment (AndyMcAliley) [reply 4_evaluate/src/interpolate_predictions.py:199]: Haha, maybe too tricky... but it's a working solution!

File: 4_evaluate/src/interpolate_predictions.py (hunk 3 above)
Comment (jdiaz4302) [reply 4_evaluate/src/interpolate_predictions.py:144]: Ah, okay. I see, and this checks out. 

broadcastingcheck

**Review by jdiaz4302 — APPROVED** [review]

File: 4_evaluate/src/interpolate_predictions.py (hunk 3 above)
Comment (AndyMcAliley) [reply 4_evaluate/src/interpolate_predictions.py:144]: Nice explanatory demo!
